clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -494,10 +580,37 @@ private void handleDisconnections(List<ClientResponse> responses, long now) {
     /**
      * Record any newly completed connections
      */
-    private void handleConnections() {
+    private void handleConnections(long now) {
         for (String node : this.selector.connected()) {
             log.debug("Completed connection to node {}", node);
-            this.connectionStates.connected(node);
+            // Though the node is connected, we might not still be able to send requests. For instance,
+            // in case of SSL connection, SSL handshake happens after connection is established.
+            if (this.requiredApiVersions == null) {
+                this.connectionStates.ready(node);
+            } else {
+                this.connectionStates.checkingApiVersions(node);
+                if (!apiVersionsRequests.contains(node))
+                    apiVersionsRequests.add(node);
+            }
+        }
+        maybeInitiateApiVersionsFetch(now);
```

Comment:
I'm wondering if we can simplify this a little bit. Once the connection has completed, it seems that both `Selector.isChannelReady()` and `InFlightRequests.canSendMore()` should both be true. So instead of going through this intermediate collection before sending the `ApiVersionRequest`, perhaps we can just immediately send the request when we transition to `CHECKING_API_VERSIONS`? I think that would let us get rid of the `apiVersionsRequests` set.

@ijuma Does that seem right?